base/mathconstants.jl

```diff
@@ -16,6 +16,26 @@ Base.@irrational γ        euler
 Base.@irrational φ        (1+sqrt(big(5)))/2
 Base.@irrational catalan  catalan
 
+const _KnownIrrational = Union{
+    typeof(π), typeof(ℯ), typeof(γ), typeof(φ), typeof(catalan)
+}
+
+function Rational{BigInt}(::_KnownIrrational)
+    Base._throw_argument_error_irrational_to_rational_bigint()
+end
+Base.@assume_effects :foldable function Rational{T}(x::_KnownIrrational) where {T<:Integer}
+    Base._irrational_to_rational(T, x)
+end
+Base.@assume_effects :foldable function (::Type{T})(x::_KnownIrrational, r::RoundingMode) where {T<:Union{Float32,Float64}}
+    Base._irrational_to_float(T, x, r)
+end
+Base.@assume_effects :foldable function rationalize(::Type{T}, x::_KnownIrrational; tol::Real=0) where {T<:Integer}
+    Base._rationalize_irrational(T, x, tol)
+end
+Base.@assume_effects :foldable function Base.lessrational(rx::Rational, x::_KnownIrrational)
+    Base._lessrational(rx, x)
+end
```

Comment:
Could this be added to the `@irrational` macro to avoid hard to debug surprises when adding more `@irrational`s in Base (users and packages outside of Base are not supposed to define `Irrational`s and use `@irrational` anyway, so it's mainly about more `@irrational`s in Base)?

Reply:
> users and packages outside of Base are not supposed to define `Irrational`s

Perhaps this was the intent, but there's nothing stopping users from doing this. Wrong effects may have catastrophic consequences like crashes or silently wrong results, or other hard-to-debug issues (undefined behavior, in short), so they're extremely undesirable.

Reply:
Sure. My only point was that maybe my suggestion is not worth it since it is supposed to be only relevant for uses in Base anyway. But even then it might be nice not to have to remember to update the union if a new `@irrational` is added.

Reply:
> Perhaps this was the intent

This intent is documented (but of course that doesn't prevent definitions outside of Base):

```julia
  │ Warning
  │
  │  This macro should not be used outside of Base Julia.
  │
  │  The macro creates a new type Irrational{:sym} regardless of where
  │  it's invoked. This can lead to conflicting definitions if two
  │  packages define an irrational number with the same name but
  │  different values.
```

Reply:
https://github.com/JuliaLang/julia/pull/47103

Reply:
The macro is irrelevant to the question of safety here. As shown in the fixed issue, it's not required to add new subtypes of `Irrational`.

Reply:
> it might be nice not to have to remember to update the union

I think there's no other way to achieve safety here.

Reply:
OK, I think there's a misunderstanding here, I'll try to clarify my initial comment:

1. The custom union only contains constants that are defined with the `@irrational` macro.
2. Instead of defining a custom union, it seems these definitions could have been added to https://github.com/JuliaLang/julia/blob/cb757c41102e4cdc899d76464acde0938c36de9e/base/irrationals.jl#L247-L257.

This wouldn't reduce the amount of code here but would ensure that for `Irrational` types defined with the `@irrational` macro (with known value etc.) these effects would be defined automatically. So if at some point an `@irrational` in Base is added, removed, or modified, it is not necessary to update the union.

Reply:
Oh, you mean to add the methods with effect annotations using the macro instead of like in this PR. Yeah, I guess that would be correct, but it'd also unfortunately increase the number of methods.

Reply:
I was wondering whether the increased number of methods would be problematic. My gut feeling was that it might be fine given that it's only five `Irrational`s currently but on the other hand I don't know the compiler internals and why/when this could be problematic (or, on the other hand, if the compiler has problems with `Union`s if they contain too many elements).